tests/scripts/check-python-files.sh

```diff
@@ -9,4 +9,15 @@
 # Run 'pylint' on Python files for programming errors and helps enforcing
 # PEP8 coding standards.
 
-pylint3 -j 2 scripts/*.py tests/scripts/*.py
+# Find the installed version of Pylint. Installed as a distro package this can
+# be pylint3 and as a PEP egg, pylint. We prefer pylint over pylint3
```

Comment:
[It should be the other way round](https://github.com/ARMmbed/mbedtls/issues/3111), because `pylint` might be Python 2. I'm not convinced that it's more common for `pylint` to be newer than `pylint3` outside of macOS, and even if it is, there's no particular reason to prefer the most recent version.

Reply:
The difference seems to be if you installed it as a Linux package you get `pylint3` and if from PEP or install it from the project site you get `pylint`.

The version of `pylint` you get seems to be independent of the naming, and actually the distro packages `pylint3`, will not necessarily be more recent that if installed from PEP or some other means.

The `3` doesn't seem to imply anything, and the fact `pylint3` is installed seems to be some historical accident.

I'm not sure why you're mentioning MacOS. That seems a red herring.

Reply:
If both `pylint3` and `pylint` are available, we can be sure that `pylint3` is suitable, but we can't be sure that `pylint` is suitable. On at least one major operating system family (Ubuntu current LTS, current Debian stable, and other Debian derivatives whose release is more than a few months old) `pylint` is not suitable.

Reply:
> we can't be sure that `pylint` is suitable

What do you mean by `pylint` isn't suitable? That sounds like you want to avoid #3087, 'Pylint fails with development branch and 2.16 branch', which sounds like sweeping a bug under the carpet than fixing it. But you don't mean that do you?

Reply:
On the current Debian stable and the current Ubuntu LTS, `pylint` is not suitable because it's for Python 2. This is unrelated to #3087, which I do want to fix.

Reply:
Ok - I think we're getting somewhere now.

`all.sh` doesn't run on Debian, and the Debian Dockerfiles are not designed to support it, but instead are designed to replace the Buildbot CI which did not run `all.sh`.

`all.sh` also no longer runs on Ubuntu 18.04 if you install `pylint` with the APT package manager rather with `pip`. That's because there's a versioning problem introduced in the dependencies of `pylint`. The Dockerfile addresses this by installing it via `pip`.

If we support the APT package of `pylint` with`all.sh`, you will break some other aspect of `all.sh`, because of the versioning issues.

To ask for support for `all.sh` on Debian is to ask for something I don't believe we've ever supported before in what is in essence, a CI script file.

Both of these points may be irrelevant to the point we came in on though - which is best, preferring `pylint`or `pylint3`?

Reply:
b`all.sh` should pass on Ubuntu 18.04. It currently doesn't because of #3087. It also currently doesn't, and won't even if #3087 is fixed, _if the `pylint` package is installed_, and this second problem is due to the change here.

I can't think of a strong reason to prefer `pylint` over `pylint3`. Yes, `pylint` _may_ be more recent than `pylint3`, but that's not necessarily the case. It makes sense to prefer a more recent version of a checking tool which may find more bugs, but it also makes sense to prefer a more established version whose behavior we understand better.

We know that `pylint3` can't be _wrong_. It may be suboptimal, if `pylint` is a “better” version of Pylint, but we aren't even sure of that. So we should favor `pylint3`. I'd also accept a patch that checks whether `pylint` is suitable and ignores it if it isn't  But I would have rejected this pull request and I intend to fix #3111 soon.

Reply:
> `all.sh` should pass on Ubuntu 18.04.

And it does pass - with the correct configuration. Install `pylint` from `pip` not from `apt-get`.

I think swapping from preferring `pylint` to `pylint3` is suboptimal, and to be honest it sounds like you fixing the scripts to fit your personal development system.

I'm not convinced there is a real issue here.

Reply:
I'm not trying to fit the scripts to my personal system, I'm trying to fit them to one the most widely used Linux distributions (Ubuntu LTS), which is used in the CI. Why would you not use the system pylint anyway?

Reply:
> Why would you not use the system pylint anyway?

This is getting circular Gilles because I already said it above. There is a conflict in the versioning of the dependencies if you install `pylint` from `apt-get`. The fact you do not have that issue implies you are not running some other aspect of `all.sh` which is likely optional.

Ubuntu 18.04 works fine with `all.sh` *if* you don't install `pylint` from `apt-get`. *If* you do, you will hit other issues.

The `ubuntu-18.04` Dockerfile which installs `pylint` from `pip` executes `all.sh` in full. Does your development system?

Reply:
What “other issues” are you talking about?

It's been a while since I ran all of `all.sh` — since it's been split into functions I only run the parts that I'm focusing on. But if any part of it fails, it would have something I haven't run in a long time.

Reply:
As stated in PR #130, it turns out installing `pylint` as a Debian package instead of as a `pip` package causes the following error in the generating the Ubuntu 18.04 Dockerfile:
```
Step 49/59 : RUN pip install yotta matplotlib > /dev/null
 ---> Running in f4150271c2a6
ERROR: pyopenssl 19.1.0 has requirement cryptography>=2.8, but you'll have cryptography 2.1.4 which is incompatible.
```
Swapping `pylint` to a `pip` package fixes this.

So you can have `pylint` or `yotta`, but you can't have both. I'm sure you don't care about `yotta` very much, but it is still supported in `mbedtls-2.16`.

There may (or may not be) be other issues. I didn't explore this in great detail as it was a very tedious distraction.

I think the key thing here is if you install python packages from `pip` and from `apt-get` it's possible to get contradictions in the versions of dependencies that cannot be resolved, so better to choose one package manager and stick by it.

Reply:
Ah, yotta through pip with pylint through apt. I don't think I've ever tried it on 18.04, only on 16.04. The Right Way is probably to install yotta in its own virtualenv, but as long as we have a working Dockerfile, I don't mind.

https://github.com/ARMmbed/mbedtls/pull/3118 should make both of us happy anyway.